### PR TITLE
feat: shared slash-command registry with Telegram menu

### PR DESCRIPTION
**Pull request by prime-axiom**
## Problem

Slash commands were duplicated between the web chat WebSocket layer and the Telegram bot — `/new` and `/stop` were hard-coded twice with slightly different wording, and Telegram had no native command menu so users had to remember the names. There was no `/help`, no read-only introspection (`/tasks`, `/cronjobs`, `/settings`), and no central place to add new commands.

## Change

A small first-version slash-command system shared by both surfaces.

- **Core (`@axiom/core`)** — new `SlashCommandRegistry` and `parseSlashCommand`. Commands carry typed metadata (`name`, `aliases`, `description`, `usage`, `surfaces`) and an optional handler. Surface-owned commands (`/new`, `/stop`, `/kill`, `/start`) register as metadata-only so they appear in `/help` and the Telegram menu while their behaviour stays inline in the surface that owns the transport state.
- **Built-in commands** — `/help`, `/tasks`, `/cronjobs` (alias `/cron`), `/settings` (alias `/model`), and `/thinking` (show/set the global main-agent thinking level). Both surfaces.
- **Telegram bot** — calls `bot.api.setMyCommands()` on start, populating the native command menu (the `/` button) from the registry. Failures are non-fatal: typed commands still work, the menu just stays empty until the next successful start.
- **Web chat** — `ws-chat.ts` intercepts typed slash commands via the registry before forwarding to the LLM; unknown commands respond with a hint pointing at `/help`. `//` (double-slash) bypass for literal slashes is preserved.
- **Docs** — `docs/web-ui/chat.md` gets a "Slash commands" section; `docs/guide/telegram.md` updates the bot-commands table to reflect the shared vocabulary and the native menu.

Destructive operations (`/undo`, `/rollback`, `/compress`) are intentionally out of scope for v1 — none have safe primitives in the codebase yet.

## Architecture notes

- A command's `handler` is optional. When omitted, the registry returns `{ kind: 'external' }` from `dispatch()` and the surface keeps full control of the response. This is how `/new` and `/stop` keep their existing behaviour (aborting the active stream, resetting the session) while still appearing in `/help`.
- The parser is conservative: only `[a-zA-Z0-9_]+` after a single leading slash, Telegram `@botname` suffix stripped, double-slash treated as escape so users can still send literal `/foo` to the LLM.
- Inspired by [Hermes Agent](https://hermes-agent.nousresearch.com/docs/reference/slash-commands)'s central `COMMAND_REGISTRY` driving both CLI and messaging surfaces, and OpenClaw's `setMyCommands`-based Telegram menu. Skill-based dynamic commands and CLI-style autocomplete are deliberately left for a later iteration.

## Testing

- CI `Lint + Baseline + Critical Flows` and `Dead code and dependency analysis` are green on the current head. Local targeted verification passed: `npm test -- packages/core/src/slash-commands.test.ts packages/telegram/src/bot.test.ts packages/web-backend/src/ws-chat.test.ts`, plus `npx eslint packages/telegram/src/bot.ts packages/telegram/src/bot.test.ts`.
- 28 slash-command tests in `packages/core/src/slash-commands.test.ts` covering parser edge cases (Telegram `@bot` suffix, double-slash escape, invalid names), registry registration/aliases/conflicts, dispatch outcomes, `/thinking`, and built-in command formatting.
- 2 new ws-chat tests for `/help` listing and unknown-command handling, plus the existing `/new` and `/stop` tests still pass unchanged.
- 3 new Telegram bot tests for `/help`, `/cron`, and `/settings` dispatching through the registry; the existing `/start` test was updated to reflect the slimmer welcome message that points at `/help`.

### CODE REVIEW

**Comment by prime-axiom** [comment]
I’m looking into the failed dead-code check now and will push a fix shortly.

**Comment by prime-axiom** [comment]
Pushed a focused CI fix for the dead-code job.\n\nWhat I changed:\n- Added  above three cross-workspace TelegramBot API methods in  (, , ).\n\nLocal verification:\n-  ✅\n-  ✅

**Comment by prime-axiom** [comment]
Dead-code CI is fixed now. The previous suppressions used the plural issue name; Fallow's suppression key is singular `unused-class-member`. Pushed `9aaa318`, and both required checks are green again.

**Comment by prime-axiom** [comment]
Added the missing `/thinking` command. It now shows or sets the main chat thinking level from both web and Telegram, persists `settings.thinkingLevel`, applies the live AgentCore update on both surfaces, and is included in the Telegram command menu via the shared registry.\n\nVerification run:\n- `npm run test -- packages/core/src/slash-commands.test.ts`\n- `npx eslint packages/core/src/slash-commands.ts packages/core/src/slash-commands.test.ts packages/web-backend/src/ws-chat.ts packages/telegram/src/bot.ts`\n\nNote: local Fallow still reports the known cross-workspace Telegram suppressions as stale in this warm workspace, while CI previously required them in a clean workspace. I left them unchanged and will verify the GitHub check result after this push.

**Comment by prime-axiom** [comment]
Checklist passed after the latest commit (73e834e). I fixed the missing `/thinking` docs and the Telegram `/cron` alias dispatch, verified the focused tests and lint for the touched files, and CI is green. GitHub does not allow me to approve my own PR, so this comment is the merge-ready signal.

**Comment by prime-axiom** [comment]
Rebased onto current `upstream/main` (`3e758c6`, fix(web-backend): inactivity watchdog) and force-pushed.

- Previous head: `73e834e`
- New head: `5aa2330`
- New base: `3e758c6` (was `0262dc7`)

### Conflict resolution

No content conflicts on rebase. After replaying onto upstream, `tsc -b` for `@axiom/telegram` started complaining about three new `bot.test.ts` mock contexts that omitted `message_id` (upstream tightened the inferred context literal type, so spread overrides without `message_id` no longer satisfy it). Added a small follow-up commit `5aa2330` (`test(telegram): include message_id in slash-command mock contexts`) that adds `message_id: 1` to the `/help`, `/cron`, and `/settings` mocks. Behaviour is unchanged.

### Verification (post-rebase, locally)

| Check | Result |
| --- | --- |
| `npm run lint` | 2 errors — pre-existing on `upstream/main` (`packages/core/src/task-runner.test.ts:1728`, `packages/core/src/task-tools.test.ts:101`, both empty `catch`); none introduced by this PR |
| `npm run lint:boundaries` | clean |
| `npm run baseline:unit` | 51 files / **1038 tests passed** |
| `npm run test -- packages/core/src/slash-commands.test.ts packages/web-backend/src/ws-chat.test.ts packages/telegram/src/bot.test.ts` | 83/83 passed |
| `npm run -w @axiom/core build`, `@axiom/telegram`, `@axiom/web-backend` | all clean |

### CI

- `Lint + Baseline + Critical Flows` ✅ SUCCESS
- `Dead code and dependency analysis` ✅ SUCCESS

`gh pr view 63` reports `mergeable: MERGEABLE`, `mergeStateStatus: CLEAN`. Merge-ready from the author's side.

**Comment by prime-axiom** [comment]
CI fix: dropped unused `export` on `handleModelCommand` in `packages/core/src/slash-commands.ts` (only consumed inside the same module). New head: `47da2f0`. Both checks green.

**Comment by prime-axiom** [comment]
## Review (full re-pass after the 5 push-ups + CI fix `47da2f0`)

### Verdict: request changes

The slash-command refactor itself looks solid (token-based callbacks with TTL + one-shot consumption, auth check on every Telegram callback, in-place message edit fallback, picker forwarded as a `system` ws message with a graceful text fallback for old clients/history reloads). Tests cover the new behavior on all three surfaces. CI is green.

The blocker is the user-facing docs.

### Blockers

1. **`docs/web-ui/chat.md`** (slash-command table, ~L118) still lists:

   ```
   | `/settings` | Show the active provider and model. (alias `/model`)                        |
   ```

   But commit `8a170cf` explicitly **removes** `/settings` and renames the primary command to `/model` (with `/provider` as the new alias). It also changes the semantics: it's no longer a read-only "show active" command — it's an interactive picker that writes `providers.json`. The table needs:
   - drop `/settings`
   - add `/model` with the new "Show or switch the active provider and model" wording (alias `/provider`)
   - mention that on the web chat the response renders as a button group (provider picker → model picker → confirmation)

2. **`docs/guide/telegram.md`** (bot-commands table, ~L77) has the same stale `/settings` row. Needs the same fix, plus a short note that on Telegram the picker renders as an inline keyboard (and that the menu auto-expires after 10 min if untouched, matching `PICKER_TOKEN_TTL_MS` — relevant for users who scroll back to an old picker).

### Non-blockers (minor / informational)

- `packages/core/src/slash-commands.ts` L246 — usage line `'/thinking <off|minimal|low|medium|high|xhigh>'` has a stray leading-space indent (two spaces instead of four), unrelated to this PR. Not worth a separate commit.
- `packages/telegram/src/bot.ts` `pickerTokenStore` is per-bot-instance, not per-user. In practice that's fine (8-hex-char tokens, 10-min TTL, auth check on resolve), but worth keeping in mind if multi-user load grows.

### Checklist summary

- Scope & Intent ✅ — single coherent refactor + new picker UI on all surfaces.
- Correctness ✅ — picker steps, case-insensitive lookup, expired-token handling all covered.
- Tests ✅ — `slash-commands.test.ts`, `bot.test.ts`, `ws-chat.test.ts` updated.
- Code Quality ✅ — clear types (`SlashCommandReply`, `SlashCommandPicker`), good "why" comments.
- Dependencies & Build ✅ — no new deps, CI green after `47da2f0`.
- Security & Safety ✅ — auth check on picker callback, one-shot tokens, TTL pruning, no secrets.
- **Docs & User-Facing ❌** — see blockers above.

**Comment by prime-axiom** [comment]
Docs blockers from review comment https://github.com/meteyou/axiom/pull/63#issuecomment-4376947687 are addressed in `eba795b`:

- `docs/web-ui/chat.md`: dropped the stale `/settings` row, added `/model` (alias `/provider`) with the new "Show or switch the active provider and model" wording, plus a short note that the response renders as an interactive button group (provider picker → model picker → confirmation) and writes `providers.json`.
- `docs/guide/telegram.md`: same swap, plus a note that on Telegram the picker renders as an inline keyboard edited in place between steps and auto-expires after 10 min if untouched (matches `PICKER_TOKEN_TTL_MS`).

No code/test changes; `npm run lint` (still only the two pre-existing baseline errors in `task-runner.test.ts:1728` / `task-tools.test.ts:101` from upstream), `npm run lint:boundaries`, and `npm run docs:build` all pass locally.

CI on `eba795b`:
- Lint + Baseline + Critical Flows: ✅ pass (1m11s) — https://github.com/meteyou/axiom/actions/runs/25362970702/job/74366687115
- Dead code and dependency analysis: ✅ pass (33s) — https://github.com/meteyou/axiom/actions/runs/25362970731/job/74366687023
- `mergeStateStatus`: CLEAN

Skipped the L246 whitespace nit on `packages/core/src/slash-commands.ts` per the review's own "not worth a separate commit" note.

Docs blockers resolved. Full checklist passes (Scope ✅, Correctness ✅, Tests ✅, Code Quality ✅, Dependencies & Build ✅, Security ✅, Docs & User-Facing ✅). Merge-ready from author's side. Not approving via `gh pr review --approve` because GitHub blocks self-approval on own PRs.

**Review by prime-axiom — COMMENTED** [review]
## Review of the two new commits (`2678213`, `c9f4258`)

### Verdict: changes requested — docs gap (blocker). Posted as a `comment` review because GitHub blocks request-changes on the PR author's own PR (`prime-axiom`); Stefan, please treat this as a hard block until the docs row below is added.

The earlier commits on this PR were already verified merge-ready in the previous round. This pass only re-runs the checklist against the two commits Stefan just pushed.

#### Findings per checklist
- **Scope & Intent** — `/tts` toggle is in scope (new Telegram-surface slash command on top of the registry). The backend startup fix is slightly off the original PR scope but small and surgical; not blocking on that alone.
- **Correctness** — `/tts` handler reads `telegram.json`, toggles `sendVoiceReply`, writes back via the same `getConfigDir()` + `JSON.stringify(...) + '\n'` pattern used by `writeConfiguredThinkingLevel`. `isHandledCommand` was extended to include `tts|voice` so the agent dispatcher does not treat them as plain messages. Backend fix wraps `getApiKeyForProvider(provider)` with `.catch()` returning `provider.apiKey || 'no-key'`, lets `agentCore` instantiate, then `restartTelegramBot()` runs — Telegram now starts even if API-key resolution throws. Reasonable.
- **Tests** — `/tts` has a focused test covering the toggle path and the on-disk write. ✅ The backend startup-despite-auth-errors path has no dedicated test (no regression guard for the `.catch()` branch). Nice to have, not a blocker.
- **Code Quality** — Clean. Aliases array, descriptive description, persistence helper kept private to the file.
- **Dependencies & Build** — `npm run lint` clean except for the two known pre-existing baseline errors in `task-runner.test.ts:1728` and `task-tools.test.ts:101` (no new errors). `npm run lint:boundaries` clean. `vitest run packages/telegram/src/bot.test.ts` 54/54 passing. `vitest run packages/web-backend/src/bootstrap` 3/3 passing. CI on `c9f4258` is green.
- **Security & Safety** — `/tts` writes only to `telegram.json` under `getConfigDir()`; no new external input or path-traversal surface. The auth-error fallback (`'no-key'`) is a sentinel string passed to the model client; downstream calls will surface a 401 the user already would have seen — it does not silently mask broken auth, just defers it past Telegram boot. OK.
- **Docs & User-Facing** — ❌ **blocker.** `/tts` (alias `/voice`) is a new user-visible Telegram slash command but is not listed in the slash-command table in `docs/guide/telegram.md`, and the `Voice messages` section does not mention the slash command alternative to the Settings → Telegram toggle. Suggestion inline at line 78. Optional polish for the `Voice messages` section (line 98) — feel free to ignore if you prefer to keep that paragraph Settings-only:

```
- **Text-to-Speech (optional)** — pick a provider and voice under [Settings → Text-to-Speech](../settings/text-to-speech), then enable [Send voice reply](../settings/telegram#send-voice-reply) on the Telegram settings page or run `/tts` (alias `/voice`) in the Telegram chat. Each agent reply is then also synthesized and uploaded as a voice/audio message.
```

### Required to unblock
Add `/tts` to the slash-command table in `docs/guide/telegram.md` (inline suggestion below). After that, this is merge-ready from my side.

File: docs/guide/telegram.md [hunk 1]
```diff
@@ -60,15 +60,35 @@ From now on, anything you send to the bot is processed by the agent. Anyone else
...
+| Command     | What it does                                                                |
+|-------------|-----------------------------------------------------------------------------|
+| `/start`    | Welcome message. Registers you as `pending` on first use.                   |
+| `/help`     | List all available slash commands.                                          |
+| `/new`      | Summarize the current session, persist it, and start a fresh conversation.  |
+| `/stop`     | Abort the current agent turn and clear queued messages. (alias `/kill`)     |
+| `/tasks`    | Show running and recent background tasks (read-only).                       |
+| `/cronjobs` | Show configured cronjobs and their next run time. (alias `/cron`)           |
+| `/model`    | Show or switch the active provider and model. (alias `/provider`)           |
+| `/thinking` | Show or set the global main-agent thinking level.                           |
```
Comment (prime-axiom) [inline docs/guide/telegram.md:78]: **blocker — Docs gate.** `feat(telegram): add voice reply toggle command` (2678213) introduces a new user-visible Telegram slash command `/tts` (with alias `/voice`) that toggles `sendVoiceReply` in `telegram.json`. The slash-command table here must list it for parity with the `setMyCommands` menu the bot now publishes.

```suggestion
| `/thinking` | Show or set the global main-agent thinking level.                           |
| `/tts`      | Toggle automatic Telegram voice replies. (alias `/voice`)                   |
```

**Comment by prime-axiom** [comment]
Docs blocker from review https://github.com/meteyou/axiom/pull/63#pullrequestreview-PRR_kwDORySLxc78z6uP is addressed in `56c00df`:

- `docs/guide/telegram.md`:
  - Slash-command table (~L79): added `/tts` row with the description `Toggle automatic Telegram voice replies. (alias /voice)`.
  - Voice messages section (~L98): extended the `Text-to-Speech (optional)` bullet to mention the in-chat alternative (`/tts`, alias `/voice`) to the Settings → Telegram **Send voice reply** toggle, including the `on`/`off`/`status` arguments and the note that the toggle persists to `telegram.json` and is independent of the global TTS settings.

`docs/web-ui/chat.md` was deliberately not touched: `/tts` is registered with `surfaces: ['telegram']` only, so it must not appear in the web slash-command table.

Local verification on `56c00df`:
- `npm run lint` — 2 errors, both pre-existing baseline on upstream (`packages/core/src/task-runner.test.ts:1728`, `packages/core/src/task-tools.test.ts:101`, both empty `catch`); none introduced by this PR.
- `npm run lint:boundaries` — clean.
- `npm run docs:build` — clean (vitepress 1.6.4, build complete in 7.33s).

CI on the new head `56c00df`:
- `Lint + Baseline + Critical Flows` ✅ SUCCESS — https://github.com/meteyou/axiom/actions/runs/25500719822/job/74832531942
- `Dead code and dependency analysis` ✅ SUCCESS — https://github.com/meteyou/axiom/actions/runs/25500719818/job/74832531840

### Full checklist re-run on PR HEAD `56c00df`

- **Scope & Intent** ✅ — single docs commit, no unrelated churn.
- **Correctness** ✅ — `/tts` toggle, `/voice` alias, `on`/`off`/`status` args, persistence target (`telegram.json`), and independence from global TTS all match the `handleTelegramVoiceCommand` implementation in `packages/telegram/src/bot.ts`.
- **Tests** ✅ — code paths unchanged; `/tts` test in `packages/telegram/src/bot.test.ts` still green from the earlier commit.
- **Code Quality** ✅ — N/A (docs only).
- **Dependencies & Build** ✅ — no dep / lockfile changes; `docs:build` passes.
- **Security & Safety** ✅ — N/A (docs only).
- **Docs & User-Facing** ✅ — the previously-failing gate. `/tts` now documented in both the Telegram slash-command table and the Voice messages section.

GitHub does not allow me to approve my own PR, so this comment is the merge-ready signal. Merge-ready from author's side — leaving the merge button to Stefan.
